yew/src/virtual_dom/vtag.rs

```diff
@@ -249,36 +244,81 @@ impl VTag {
         }
     }
 
-    /// This handles patching of attributes when the keys are equal but
-    /// the values are different.
-    fn diff_attributes<'a>(
-        &'a self,
-        ancestor: &'a Option<Box<Self>>,
-    ) -> impl Iterator<Item = Patch<&'a str, &'a str>> + 'a {
-        // Only change what is necessary.
-        let to_add_or_replace =
-            self.attributes.iter().filter_map(move |(key, value)| {
-                match ancestor
-                    .as_ref()
-                    .and_then(|ancestor| ancestor.attributes.get(&**key))
-                {
-                    None => Some(Patch::Add(&**key, &**value)),
-                    Some(ancestor_value) if value != ancestor_value => {
-                        Some(Patch::Replace(&**key, &**value))
-                    }
-                    _ => None,
+    /// Diffs and patches changes in attributes between the old and new list.
+    ///
+    /// This method optimises for the lists being the same length, having the
+    /// same keys in the same order - most common case.
```

Comment:
There are 3 reasons to use vectors instead of hashmaps:
1. A use case with an if statement, as you demonstrated, is arguably less common than simply using a static set of keys in the same order.
2. The most common alteration of attributes should be adding and removing attributes. In that case the initial length check in  `patch_attributes()` will cheaply fallback to hashmap diffing. 
3. The `html` macro constructs attribute lists as vectors of tuples. Having the input format and storage format differ prevents us from doing cheap in-place construction of the attribute vector and requires for a vector to be constructed first, then converted to an iterator and only then one by one used to construct a hashmap. Additionally, such in-place construction helps the compiler produce more parallel and more eagerly inlined assembly.

Reply:
> A use case with an if statement, as you demonstrated, is arguably less common than simply using a static set of keys in the same order.

I don't disagree that using a static set of keys in the same order is the most common case. But Yew is not just optimizing for the most common case. We should try to have a consistent experience for as many reasonable use cases as possible.

Another thing to consider is that it will soon be very easy to have conditional attributes (#1433) which will disrupt the order of attributes and could be pretty common. Actually, this is already supported for `"class"` too.

> The most common alteration of attributes should be adding and removing attributes. In that case the initial length check in patch_attributes() will cheaply fallback to hashmap diffing.

Maybe we can find a way that handles both cases well? If we use an `IndexMap` we could iterate over each attribute map in order and fall back to lookups without an extra allocation

> The html macro constructs attribute lists as vectors of tuples. Having the input format and storage format differ prevents us from doing cheap in-place construction of the attribute vector and requires for a vector to be constructed first, then converted to an iterator and only then one by one used to construct a hashmap. Additionally, such in-place construction helps the compiler produce more parallel and more eagerly inlined assembly.

Good points. Hmm... what do you think about creating a hashmap in the macro instead? The macro could generate code like `HashMap::collect(&[("class", "btn"), ("href", "...")])`.

Reply:
It seems benchmarks are required. I'll do that.

Reply:
We don't have any micro benchmarks yet unfortunately. I imagine them being run with the wasm-bindgen test runner to imitate the actual execution environment.

Reply:
I'll look into it.

Reply:
Benchmark results: 

```
    bench_diff_attributes_change_last: hashmaps:         4ns (R²=0.982, 201750 iterations in 103 samples)
    bench_diff_attributes_change_last: vectors:         1ns (R²=0.988, 696525 iterations in 116 samples)
    bench_diff_attributes_change_middle: hashmaps:         4ns (R²=0.998, 221926 iterations in 104 samples)
    bench_diff_attributes_change_middle: vectors:         1ns (R²=0.992, 696525 iterations in 116 samples)
    bench_diff_attributes_change_first: hashmaps:         4ns (R²=0.926, 201750 iterations in 103 samples)
    bench_diff_attributes_change_first: vectors:         1ns (R²=0.997, 633203 iterations in 115 samples)
    bench_diff_attributes_append: hashmaps:         4ns (R²=0.998, 221926 iterations in 104 samples)
    bench_diff_attributes_append: vectors:         4ns (R²=0.997, 221926 iterations in 104 samples)
    bench_diff_attributes_same: hashmaps:         3ns (R²=0.981, 268533 iterations in 106 samples)
    bench_diff_attributes_same: vectors:          0s (R²=0.995, 1493077 iterations in 124 samples)
```

Predictably, comparing the same list or a list changed in the middle/end is faster, when storing attributes as vectors. What is puzzling is the first attribute changing still being notably faster. Probably due to branch prediction. Branch hinting is still nightly-only, so nothing can be done about that.

Reply:
Predictably worse than hashmaps.
```
    bench_diff_attributes_change_last: hashmaps:         3ns (R²=0.991, 221926 iterations in 104 samples)
    bench_diff_attributes_change_last: indexmap:         4ns (R²=1.000, 221926 iterations in 104 samples)
    bench_diff_attributes_change_last: vectors:          0s (R²=0.994, 766179 iterations in 117 samples)
    bench_diff_attributes_change_middle: hashmaps:         3ns (R²=0.999, 244120 iterations in 105 samples)
    bench_diff_attributes_change_middle: indexmap:         4ns (R²=0.998, 244120 iterations in 105 samples)
    bench_diff_attributes_change_middle: vectors:          0s (R²=0.999, 842798 iterations in 118 samples)
    bench_diff_attributes_change_first: hashmaps:         3ns (R²=0.999, 244120 iterations in 105 samples)
    bench_diff_attributes_change_first: indexmap:         4ns (R²=0.999, 244120 iterations in 105 samples)
    bench_diff_attributes_change_first: vectors:          0s (R²=0.999, 766179 iterations in 117 samples)
    bench_diff_attributes_append: hashmaps:         3ns (R²=0.999, 244120 iterations in 105 samples)
    bench_diff_attributes_append: indexmap:         4ns (R²=0.999, 221926 iterations in 104 samples)
    bench_diff_attributes_append: vectors:         3ns (R²=0.997, 244120 iterations in 105 samples)
    bench_diff_attributes_same: hashmaps:         3ns (R²=0.997, 295388 iterations in 107 samples)
    bench_diff_attributes_same: indexmap:         3ns (R²=0.998, 268533 iterations in 106 samples)
    bench_diff_attributes_same: vectors:          0s (R²=0.994, 1642386 iterations in 125 samples)
```

Reply:
I added a common on the indexmap benchmark. I expect it to perform better if you optimize for the common case where attributes are in the same order

Reply:
@bakape looks like `IndexMap` is > 50% perf improvement over `HashMap` when optimized. I also improved the vector algorithm by using the linear diff even when the attribute length is unequal.

```
    bench_diff_attributes_change_last: hashmaps:         3ns (R²=0.990, 268533 iterations in 106 samples)
    bench_diff_attributes_change_last: indexmap:         2ns (R²=0.999, 475732 iterations in 112 samples)
    bench_diff_attributes_change_last: vectors:          0s (R²=0.999, 1121768 iterations in 121 samples)
    bench_diff_attributes_change_middle: hashmaps:         3ns (R²=0.999, 324928 iterations in 108 samples)
    bench_diff_attributes_change_middle: indexmap:         2ns (R²=0.997, 475732 iterations in 112 samples)
    bench_diff_attributes_change_middle: vectors:          0s (R²=0.999, 1121768 iterations in 121 samples)
    bench_diff_attributes_change_first: hashmaps:         3ns (R²=0.999, 324928 iterations in 108 samples)
    bench_diff_attributes_change_first: indexmap:         2ns (R²=0.999, 475732 iterations in 112 samples)
    bench_diff_attributes_change_first: vectors:          0s (R²=0.999, 1019788 iterations in 120 samples)
    bench_diff_attributes_append: hashmaps:         3ns (R²=0.997, 295388 iterations in 107 samples)
    bench_diff_attributes_append: indexmap:         1ns (R²=0.999, 633203 iterations in 115 samples)
    bench_diff_attributes_append: vectors:         1ns (R²=0.999, 927079 iterations in 119 samples)
    bench_diff_attributes_same: hashmaps:         2ns (R²=0.999, 393165 iterations in 110 samples)
    bench_diff_attributes_same: indexmap:         1ns (R²=0.999, 633203 iterations in 115 samples)
    bench_diff_attributes_same: vectors:          0s (R²=0.999, 2645086 iterations in 130 samples)
```

Reply:
Here's the indexmap code I used:
```rust
fn diff_attributes_indexmap<'a>(
    new: &'a IndexMap<&'static str, Cow<'static, str>>,
    old: &'a IndexMap<&'static str, Cow<'static, str>>,
) -> Vec<Patch<&'static str, &'a str>> {
    let mut out = Vec::new();
    let mut new_iter = new.iter().peekable();
    let mut old_iter = old.iter().peekable();
    loop {
        if new_iter.peek().is_none() || old_iter.peek().is_none() { break }
        match (new_iter.next(), old_iter.next()) {
            (Some(n), Some(o)) => {
                if n.0 != o.0 { break }
                if n.1 != o.1 {
                    out.push(Patch::Replace(*n.0, n.1.as_ref()));
                }
            }
            _ => break
        }

    }
    diff_incompatible_indexmap(&mut out, new_iter, old_iter, &new, &old);
    return out;
}

fn diff_incompatible_indexmap<'a>(
    dst: &mut Vec<Patch<&'static str, &'a str>>,
    new_keys: Peekable<Iter<'a, &'static str, Cow<'static, str>>>,
    old_keys: Peekable<Iter<'a, &'static str, Cow<'static, str>>>,
    new_map: &'a IndexMap<&'static str, Cow<'static, str>>,
    old_map: &'a IndexMap<&'static str, Cow<'static, str>>,
) {
    for (k, n_v) in new_keys {
        match old_map.get(k) {
            Some(o_v) => {
                if n_v != o_v {
                    dst.push(Patch::Replace(k, n_v));
                }
            }
            None => dst.push(Patch::Add(k, n_v)),
        };
    }

    for (k, _) in old_keys {
        if !new_map.contains_key(k) {
            dst.push(Patch::Remove(k));
        }
    }
}
```